### ISSUE TITLE
changing name of project in composer.json in fork/branch impossible?

### ISSUE THREAD

**Issue by matthias-chlechowitz** [composer/composer#4677]
hi everybody,

i currently ran into a problem with renaming a project in composer.json.

the previos status was like the following:

```
{
    "name": "old/name",
    "license": "Business",
    "type": "library",
    "description": "description",
    "autoload": {},
    "minimum-stability":"dev",
    "require": {
        "php": ">=5.3.3"
    },
    "require-dev": {
      "sensio/generator-bundle": "~2.5",
      "phpunit/phpunit": "~4.6",
      "mockery/mockery": "0.9.4",
      "leaphub/phpcs-symfony2-standard": "~2.0.0"
    },
    "scripts": {
        "post-root-package-install": [
        ],
        "post-install-cmd": [
        ],
        "post-update-cmd": [
        ],
        "prepare-tests": [
        ]
    },
    "config": {
        "bin-dir": "bin"
    },
    "extra": {
        },
        "branch-alias": {
            "dev-master": "2.6-dev",
        }
    }
}
```

now i need to change this to 
```
{
    "name": "new/name",
    "license": "Business",
    "type": "library",
    "description": "description",
....
}
```

therefore i created a new branch, changed the information, created a branch-alias to reference it in the target-project.

```
...
    "repositories": [
        {
            "type": "vcs",
            "url": "git@github.com:matthias-chlechowitz/forked-repo.git"
        }
    ],
    "require": {
        "new/name":"dev-matthias/rewrite-branchname"
        }
...
```

but when running composer update to fetch the updated version, the old version with the previous project name is fetched and the following error message appears:

```
  Problem 1
    - The requested package cr/commons could not be found in any version, there may be a typo in the package name.
```

so it's not possible to rename projects within one repo or even forks? or am i missing a major part?

best regards,
matthias

**Comment by alcohol** [comment]
You will need to provide more info, such as the name of the branch etc.

**Comment by stof** [comment]
Composer reads the package name from the main branch of the repo (the main branch is either the branch configured as main one on github, or ``master`` for repos hosted elsewhere)

**Comment by matthias-chlechowitz** [comment]
@alcohol i just wanted to try to reproduce it with public repos, but the answer of @stof makes it unnecessary. thanks a lot for the clarification.

**Comment by murnieza** [comment]
Maybe it makes sense to change this behaviour and read required branch instead of default one?

**Comment by modelm** [comment]
I agree with @murnieza. If I require a package and specify a branch, I expect composer to define the package based on the composer.json in that branch, not the default branch. This is especially problematic with VCS packages when forking a repository I don't have push access to in order to work on a feature branch - my fork has to merge the feature branch into the default branch for composer to even recognize the correct package name, which kind of defeats the purpose.

**Comment by Seldaek** [comment]
You don't *need* to change the name to require your feature branch though, so I don't see the problem here.

**Comment by modelm** [comment]
I do because my github username is not the same as the user from whom I forked the repo.

e.g.

original-user/repo-name -> my-user/repo-name

if I am using any branch other than master, this just makes composer fail to find the package until master is updated to match.

**Comment by Seldaek** [comment]
No you don't. There is no link between github usernames and package names. All you have to do is add a [VCS repository](https://getcomposer.org/doc/05-repositories.md#vcs) and then require your branch using `dev-<feature branch name>`.

**Comment by modelm** [comment]
It was/is already a VCS respository, under the old name, using a non-default branch.

When I changed the VCS definition in the parent's composer.json to point to the new fork, composer failed to find it because the package name on the package's master branch composer.json still had the old username. It was not until I updated that composer.json (by merging the feature branch) that composer finally identified the package, since by this point the master branch now defines the package according to the new fork.

**Comment by Seldaek** [comment]
Well that just sounds incorrect to me. I am pretty sure this works as advertised so either you hit a bug or you did it wrong somewhere along the line. Renaming is **not** necessary for using a fork. Please try it again and if you really encounter an issue create a new issue with full repro details.

**Comment by modelm** [comment]
https://github.com/composer/composer/issues/4984

**Comment by tishma** [comment]
So, the bottom line is - it's not fork that breaks composer. It is rename. In a branch. So, may I add that the exact reason for me to fork was to fight a library's compatibility issue with an old php version, which is only supported in a branch of package I'm depending on. I feel that keeping original name would be deceiving for anyone trying to maintain the codebase (including me). I'm going for a pull request, but I still can see no reason not to support renaming package in a branch.

**Comment by Seldaek** [comment]
@tishma the problem is if we take the name in every branch/tag then renaming a package on packagist becomes kinda impossible as you would have to re-tag everything and update the name in all the old composer.json files. The way it's done right now, we take the name from the default branch and then that's the name used for everything else in that repo, so just change the name in master, or change the default branch of the fork to be your patched-branch, and it should work fine.

**Comment by tishma** [comment]
@Seldaek Interesting! Fork's default branch is probably what I want. Thanks. I wouldn't say rename in packagist sounds like a good idea, but since it's already supported - I wouldn't want to break it :)

**Comment by Seldaek** [comment]
Well rename in packagist isn't allowed, but if you rename a package after a few versions and submit it again to packagist then all versions are made available with the new name, which is a good thing to have IMO.

**Comment by trip-somers** [comment]
FWIW, this reason for not doing anything applies only to Packagist usage of Composer. Private/internal (non-Packagist) use is also affected by this.

I was working on a private repository and run into this exact issue because we needed to change a package name due to some internal naming convention changes. The RC branches weren't found by Composer because it was still seeing the old package name.
